initial hexagon parametric shape

shimwell opened the pull request:
## Proposed changes

This PR aims to add another parametric shape to the collection. This hexagon face extrusion can be useful for HCPB blankets and I guess even fission reactors pin arrangements or bee hives :-)

Feedback or code contribution are most welcome ;-)

@bruceey42

Usage example than makes the shape and exports it as a stp file

```python
import paramak
hex = paramak.HexagonPin(length_of_side=5, distance=10, center_point=(10,20))
hex.export_stp('hex10-20.stp')
```
TODO

- [x] add code to add the center points coordinates to the base hex coordinates
- [x] tests, check shape construction and volume tests and impact of center point tests
- [x] documentation (including labeled svg image)

![Screenshot from 2021-02-02 14-08-58](https://user-images.githubusercontent.com/8583900/106611743-3aa3a100-6560-11eb-8e83-631f9d2566e7.png)


## Types of changes

What types of changes does your code introduce to the Paramak?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Pep8 applied
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

shimwell committed:
initial hexagon parametric shape

shimwell committed:
added example hexagon

Commit:
Automated autopep8 fixes

RemDelaporteMathurin commented:
HCPB blankets!!!! 🎉

RemDelaporteMathurin commented on examples/example_parametric_components/make_all_parametric_components.py:
```diff
@@ -417,6 +417,13 @@ def main():
 
+    component = paramak.HexagonPin(
+        length_of_side=5,
+        distance=10,
```
I know this is an ongoing discussion but I still think "distance" isn't descriptive enough.
I find that "extrude_distance" would be clearer. But I understand that would require to be changed in all components...

Shimwell replied:
ok we can change this across the collection sometime, but perhaps not in this PR

RemDelaporteMathurin commented on paramak/parametric_components/hexagon_pin.py:
```diff
@@ -0,0 +1,96 @@
+
+    Args:
+        length_of_side: the vertical (z axis) height of the coil (cm).
+        width: the horizontal (x axis) width of the coil (cm).
```
I think width should be distance so that it's consisten with the init function

Shimwell replied:
thanks for that, issue has been fixed

RemDelaporteMathurin commented on paramak/parametric_components/hexagon_pin.py:
```diff
@@ -0,0 +1,96 @@
+
+    @property
+    def center_point(self):
+        return self._center_point
```
This line isn't tested.

Shimwell replied:
Lots more tests have now been added. There center point setting and getting is tested in these new tests. Ideally mores tests would be added but I think this component is above average now

RemDelaporteMathurin reviewed (changes requested):
Very nice addition! Can't wait to see BBs added!
There's just a docstrings inconsistence + an untested line :-D

Shimwell commented:
@bruceey42 here is a copy paste of the code you came up with earlier 

```python
        for entry in points:
            entry = (entry[0] + self.center_point[0], entry[1] + self.center_point[1])
```

bruceey42 committed:
Update hexagon_pin.py

Cleaned up the docstring and added changed the point generation to be more concise and to add the translation of the profile around the self.center_point.

Commit:
Automated autopep8 fixes

shimwell committed:
added example tests for improvment

Commit:
Automated autopep8 fixes

bruceey42 committed:
Update test_HexagonPin.py

Made changes to the tests with helpful instruction from the Jo(h)ns.

Commit:
Automated autopep8 fixes

Shimwell commented:
confirming the new tests work over here and look good. I like the loop through the areas you added at the end

Shimwell commented:
image for the docs 
![hexagon_pin(3)](https://user-images.githubusercontent.com/8583900/107091965-88d3d100-67fa-11eb-9a95-b0df0cf104f3.png)

shimwell committed:
added hex pin to docs

shimwell committed:
corrected typo

Shimwell commented:
Confirming that the docs work

@billingsley-john would you be able to go over the automated docs and the "added hex pin to docs" commit to @bruceey42 sometime next week. Thanks

![Uploading Screenshot from 2021-02-05 21-47-17.png…]()

bruceey42 committed:
Minor test_HexagonPin update

Slightly more elegant but not noticeably faster.

Commit:
Automated autopep8 fixes

Shimwell commented:
> HCPB blankets!!!! tada

I guess this is a nice start, we might need to add custom extrusion angles in the future to get all the flexibility needed for HCPB blankets

billingsley-john reviewed (approved):
All looks good!
